README.md
fusion_tables
==============
<iframe style="float:right;" width="280px" height="150px" scrolling="no"  src="http://www.google.com/fusiontables/embedviz?viz=MAP&q=select+col0%2C+col1%2C+col2%2C+col3%2C+col4%2C+col5%2C+col6%2C+col7%2C+col8%2C+col9%2C+col10%2C+col11%2C+col12%2C+col13+from+245192+&h=false&lat=51.509383501611595&lng=-0.13586997985839844&z=12&t=4&l=col11"></iframe>

This gem lets you easily interact with [Google Fusion Tables API](http://www.google.com/fusiontables/Home) from your Ruby application via a plain SQL interface, or an object orientated interface.

 

Demo and examples
------------------

* Twitter [demo](http://tables.googlelabs.com/DataSource?snapid=73106) / [code](http://github.com/tokumine/fusion_tables/blob/master/examples/compare_tweets.rb) /
[blog](http://www.tokumine.com/2010/08/10/fusion-tables-gem/)
* Boris bike [demo](http://tables.googlelabs.com/DataSource?snapid=78314) / [code](http://github.com/tokumine/fusion_tables/blob/master/examples/boris_bikes.rb) 
* [Tests](http://github.com/tokumine/fusion_tables/tree/master/test/)


Installation
-------------

``` bash
gem install fusion_tables
```

**Gem Dependencies**

* gdata_19 >= 1.1.2

**Rubies**

* 1.8.7
* 1.9.2

Usage 
------
``` ruby
require 'fusion_tables'
	
	
# Connect to service	
@ft = GData::Client::FusionTables.new      
@ft.clientlogin(username, password)


# 1. SQL interface
# =========================
@ft.execute "SHOW TABLES" 
@ft.execute "INSERT INTO #{my_table_id} (name, geo) VALUES ('tokyo', '35.6894 139.6917');"
@ft.execute "SELECT count() FROM #{my_table_id};"


# 2. ORM interface
# ========================
# Browse existing tables
@ft.show_tables
 # => [table_1, table_2] 

# Getting table id suitable for using with google maps (see more below)
table_1.id #=> 42342 (the table's google id)

# Count data
table_1.count #=> 1

# Select data
table_1.select 
 #=> data

# Select data with conditions
table_1.select "name", "WHERE x=n"
 #=> data

# Select ROWIDs
row_ids = table_1.select "ROWID"

# Drop tables
@ft.drop table_1.id                    # table id
@ft.drop [table_1.id, table_2.id]     # arrays of table ids
@ft.drop /yacht/                      # regex on table name

# Creating a table
cols = [{:name => "friend name",    :type => 'string' },
        {:name => "age",            :type => 'number' },
        {:name => "meeting time",   :type => 'datetime' },
        {:name => "where",          :type => 'location' }]

new_table = @ft.create_table "My upcoming meetings", cols

# Inserting rows (auto chunks every 500)
data = [{"friend name" 	=> "Eric Wimp", 
         "age"          => 25, 
         "meeting time" => Time.utc(2010,"aug",10,20,15,1),
         "where"        => "29 Acacia Road, Nuttytown"}]
new_table.insert data

# Currently FT API only supports single row UPDATE.
new_table.update 1, [{"friend name"	=> "Bananaman"}]

# Delete row
new_table.delete 1

# Delete all rows
new_table.truncate!
```


Known Issues
-------------

* The Google gdata_19 gem conflicts with the GData2 gem. Only current fix is to uninstall GData2.
* You have to make a table public before you can display it on a map. This can only be done via FT web interface. 

Note on Patches/Pull Requests
------------------------------
 
* Fork the project.
* Make your feature addition or bug fix.
* Add tests for it. This is important so I don't break it in a
  future version unintentionally.
* Commit, do not mess with rakefile, version, or history.
  (if you want to have your own version, that is fine but bump version in a commit by itself I can ignore when I pull)
* Send me a pull request. Bonus points for topic branches.



Contributors
-------------

Largely based on Tom Verbeure's [work for MTBGuru](http://code.google.com/p/mtbguru-fusiontables/)

* tokumine
* sikachu
* troy
* wynst
* sdball


CHANGELOG
== 0.3.1 (21/08/11)

* updated to use new HTTPS endpoints (sdball)

== 0.3.0 (05/06/11)

* simple SQL interface API

== 0.2.4 (03/06/11)

* bugfixes

== 0.2.3 (03/06/11)

* all tests pass
* readme simplified
* count now returns 0 for empty tables
* truncate table, or delete all rows (sikachu)
* fix up the single update row method (sikachu)
* if value is nil, insert empty string (wynst)
* fixes to exception handling and tests (troy)

== 0.2.2

* 1.9.2 compatibility through gdata_19 gem
* examples updated for 1.9.2
* datetime columns now include time



test/helper.rb
require 'rubygems'
require 'test/unit'
require 'shoulda'
require 'yaml'

$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
$LOAD_PATH.unshift(File.dirname(__FILE__))
require 'fusion_tables'

class Test::Unit::TestCase
    
  def init_config
    if not defined? @config_file
      begin
        @config_file = YAML::load_file(File.join(File.dirname(__FILE__), 'test_config.yml'))
      rescue
        puts "Please configure your test_config.yml file using test_config.yml.sample as base"
      end    
    end
    @config_file
  end

  def username
    @config_file['username']
  end

  def password
    @config_file['password']
  end
  
  def table_name
    @config_file['table_name']
  end  
end


test/test_ext.rb
require 'helper'

class TestExt < Test::Unit::TestCase
 
  context "The Fusion Tables helper functions" do
    setup do 
      init_config
      @ft = GData::Client::FusionTables.new      
      @ft.clientlogin(username, password)
    end
  
    
    should "raise ArgumentError if supply unknown types to it" do
      assert_raise ArgumentError do
        @ft.create_table "test table", [{:name => "test_col", :type => "billys birthday" }]
      end
    end  
  
    should "let you create a table if you get everything right" do
      table = @ft.create_table "test_table", [{:name => "test_col", :type => "string" }]
      assert_equal GData::Client::FusionTables::Table, table.class
      @ft.drop(table.id)
    end
    
    should "correct your table name to a certain degree on create" do
      table = @ft.create_table "test table", [{:name => "test col", :type => "string" }]
      assert_equal "test_table", table.name
      @ft.drop(table.id)
    end
    
    should "return you a list of your fusion tables" do
      resp = @ft.show_tables
      assert_equal GData::Client::FusionTables::Table, resp.first.class if resp.first
    end
    
    should "be possible to delete a table with an id" do
      table = @ft.create_table "test_table", [{:name => "test col", :type => "string" }]
      assert_equal 1, @ft.drop(table.id)
    end
    
    should "be possible to delete tables with an array of ids" do
      table1 = @ft.create_table "test_table", [{:name => "test col", :type => "string" }]
      table2 = @ft.create_table "test_table", [{:name => "test col", :type => "string" }]
      assert_equal 2, @ft.drop([table1.id, table2.id])
    end
     
    should "be possible to delete multiple tables with a regex" do
      table1 = @ft.create_table "test_table", [{:name => "test col", :type => "string" }]
      table2 = @ft.create_table "test_table", [{:name => "test col", :type => "string" }]
      assert_equal 2, @ft.drop(/^test_/)
    end
       
    should "return zero if passed a silly id" do
      assert_equal 0, @ft.drop(235243875629384756)
    end    
  end

end


test/test_sql.rb
require 'helper'

class TestTable < Test::Unit::TestCase
  
  context "testing plain SQL" do
    setup do 
      init_config
      @ft = GData::Client::FusionTables.new      
      @ft.clientlogin(username, password)
      @table = @ft.create_table "test", [{:name  => 'firstname',  :type => 'string'},
                                         {:name  => 'phone',      :type => 'number'},
                                         {:name  => 'dob',        :type => 'datetime'},
                                         {:name  => 'house',      :type => 'location'}]
    end
    
    should "be able to SHOW TABLES" do
      ret = @ft.execute "SHOW TABLES"      
      assert !ret.empty?
      assert_equal Array, ret.class
    end  
    
    should "be able to DESCRIBE a table" do
      ret = @ft.execute("DESCRIBE #{@table.id}")      
      expected = [{:column_id=>"col0", :name=>"firstname", :type=>"string"},
                  {:column_id=>"col1", :name=>"phone", :type=>"number"},
                  {:column_id=>"col2", :name=>"dob", :type=>"datetime"},
                  {:column_id=>"col3", :name=>"house", :type=>"location"}]
      assert_equal expected, ret                 
    end
    
    should "be able to INSERT a row" do
      ret = @ft.execute "INSERT INTO #{@table.id} (firstname) VALUES ('eric');"
      assert_equal 1, ret.first[:rowid].to_i
    end  
    
    should "be able to SELECT a row" do
      @ft.execute "INSERT INTO #{@table.id} (firstname) VALUES ('eric');"      
      ret = @ft.execute "SELECT ROWID, firstname FROM #{@table.id}"
      assert_equal [{:rowid=>"1", :firstname=>"eric"}], ret
    end  
    
    should "be able to UPDATE a row" do
      @ft.execute "INSERT INTO #{@table.id} (firstname) VALUES ('eric');"      
      
      ret = @ft.execute "SELECT ROWID, firstname FROM #{@table.id}"
      assert_equal [{:rowid=>"1", :firstname=>"eric"}], ret
      
      ret = @ft.execute "UPDATE #{@table.id} SET firstname='simon' WHERE ROWID = '#{ret.first[:rowid]}'"
      ret = @ft.execute "SELECT ROWID, firstname FROM #{@table.id}"
      assert_equal [{:rowid=>"1", :firstname=>"simon"}], ret      
    end
    
    should "be able to DELETE a row" do
      @ft.execute "INSERT INTO #{@table.id} (firstname) VALUES ('eric');"
      @ft.execute "INSERT INTO #{@table.id} (firstname) VALUES ('eric');"
            
      ret = @ft.execute "SELECT ROWID FROM #{@table.id};"      
      @ft.execute "DELETE FROM #{@table.id} WHERE ROWID = '#{ret.first[:rowid]}';"      
      
      ret = @ft.execute "SELECT count() FROM #{@table.id};"
      
      assert_equal 1, ret.first[:"count()"].to_i  
    end    
    
    # should "be able to query geographic data" do
    #   @table = @ft.create_table "test", [{:name  => 'name',      :type => 'string'},
    #                                      {:name  => 'geo',       :type => 'location'}]
    #       
    #   @ft.execute "INSERT INTO #{@table.id} (name, geo) VALUES ('tokyo',   '35.6894 139.6917');
    #                INSERT INTO #{@table.id} (name, geo) VALUES ('osaka',   '34.6937 135.5021');
    #                INSERT INTO #{@table.id} (name, geo) VALUES ('fukuoka', '33.5903 130.4017');
    #                INSERT INTO #{@table.id} (name, geo) VALUES ('kyoto',   '35.0116 135.7680');
    #                INSERT INTO #{@table.id} (name, geo) VALUES ('nagoya',  '35.1814 136.9063');"
    #   
    #   # get cities nearest to Nara
    #   res = @ft.execute "SELECT * FROM #{@table.id} ORDER BY ST_DISTANCE(geo, LATLNG(35.6894,139.6917)) LIMIT 10"
    #   puts res.inspect
    # end 
  end  
end    

examples/compare_tweets.rb
# encoding: UTF-8
#
# Twitter Fusion Tables Mashup
# S.Tokumine 2010
#
# looks for tweets in the live stream around certain 
# cities and posts them to fusion tables with KML attached
#
# Gem dependencies:
#
# tweetstream
# GeoRuby
# fusion_tables
#
# Output from running this for an evening
# http://tables.googlelabs.com/DataSource?snapid=72509
#
require 'rubygems'
require 'tweetstream'
require 'geo_ruby'
include GeoRuby
include SimpleFeatures
require 'fusion_tables'
require 'time'
require 'yaml'

class Object
  def try(method, *args, &block)
    send(method, *args, &block)
  end
end

# Configure settings
config = YAML::load_file(File.join(File.dirname(__FILE__), 'credentials.yml'))
DEFAULT_SRID = 4328

# Twitter places
places = {
  :san_francisco => [-122.75,36.8,-121.75,37.8],
  :new_york      => [-74,40,-73,41],
  :tokyo         => [139.3,35,140.3,36],
  :london        => [-0.54,51.2,0.46,52.2],
  :madrid        => [-4.2,40,-3.2,41],
  :paris         => [1.75,48.5,2.75, 49.5],
  :beijing       => [115.9,39,116.9,40],
  :mumbai        => [72.75,18.88,73.75,19.88],
}

# Configure fusion tables
ft = GData::Client::FusionTables.new
ft.clientlogin(config["google_username"], config["google_password"])
table_name = "TwitterFusion"
cols = [
  {:name => 'screen_name',  :type => 'string'},  
  {:name => 'avatar',       :type => 'string'},
  {:name => 'text',         :type => 'string'},
  {:name => 'created',      :type => 'datetime'},
  {:name => 'url',          :type => 'string'},
  {:name => 'location',     :type => 'location'},
  {:name => 'iso',          :type => 'location'},
  {:name => 'country_name', :type => 'location'},  
  {:name => 'city',         :type => 'string'}
]

# Create FT if it doesn't exist
tables = ft.show_tables
table  = tables.select{|t| t.name == table_name}.first
table  = ft.create_table(table_name, cols) if !table

# Configure Twitter stream client
data = []
tw = TweetStream::Client.new(config["twitter_username"],config["twitter_password"])

# configure friendly rate limit handling
tw.on_limit do |skip_count|
  sleep 5
end 

# start searching twitter stream and posting to FT
tw.filter(:locations => places.values.join(",")) do |tweet|  
  begin    

    country = "unknown"
    iso     = "unknown"    
    begin
      country = tweet.try(:[],:place).try(:[], :country)
      iso = tweet.try(:[],:place).try(:[], :country_code)
    rescue
    end
        
    # Divine the tweets geometry
    #
    # overly complex due to 
    # * some US tweets have their lat/longs flipped (but not all...)
    # * some geo tweets are made using a "place" envelope rather than exact lat/kng      
    if tweet[:geo]
      if iso == 'US' && tweet[:geo][:coordinates][1] > 0
        p = Point.from_x_y(tweet[:geo][:coordinates][0],tweet[:geo][:coordinates][1]) 
      else
        p = Point.from_x_y(tweet[:geo][:coordinates][1],tweet[:geo][:coordinates][0]) 
      end  
    else
      p = Polygon.from_coordinates(tweet[:place][:bounding_box][:coordinates]).envelope.center
    end
  
    # work out which city the tweet is from by testing with an extended bounding box
    # BBox extention needed as twitter returns things outside our defined bboxes...
    city = "unknown"
    places.each do |key, value|
      if !(p.x < value[0]-1 || p.x > value[2]+1 || p.y < value[1]-1 || p.y > value[3]+1)         
        city = key.to_s.gsub("_"," ")    
        break
      end
    end
            
    # pack data    
    data << {
      "screen_name"  => tweet[:user][:screen_name],
      "avatar"       => tweet[:user][:profile_image_url],
      "text"         => tweet[:text],
      "created"      => Time.parse(tweet[:created_at]),
      "url"          => "http://twitter.com/#{tweet[:user][:screen_name]}/status/#{tweet[:id]}",
      "location"     => p.as_kml,
      "iso"          => iso,
      "country_name" => country,
      "city"         => city
    } 
  rescue => e
    puts "ERROR: #{e.inspect}, #{e.backtrace}"
    #let sleeping dogs lie...
  end  
  
  # let us know how we're doing
  puts "#{50-data.size}: #{city}, #{tweet.text}"
  
  # Post to fusion tables  
  if data.size == 50
    puts "sending data to fusion tables..."
    ft_data = data
    data = []
    table.insert ft_data
  end      
end

lib/fusion_tables/data/data.rb
# Copyright (C) 2010 Simon Tokumine
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

module GData
  module Client
    class FusionTables < Base
      class Data
        include Enumerable
        
        attr_reader :headers, :body, :live
        alias :to_h :body  
        
        # configures headers hash and sets up
        def initialize options
          @headers = options[:headers]
          @body = options[:body]          
        end
        
        # Reads in CSV
        def self.parse response
          body = []
          headers = []
          
          first = true
          if CSV.const_defined? :Reader
            CSV::Reader.parse(response.body) do |row|
              if first
                first = false
                headers = row.map { |x|x.strip.downcase.gsub(" ","_").to_sym }
                next
              end
              body << Hash[*headers.zip(row).flatten]
            end
          else
            CSV.parse(response.body) do |row|
              if first
                first = false
                headers = row.map { |x|x.strip.downcase.gsub(" ","_").to_sym }
                next
              end
              body << Hash[*headers.zip(row).flatten]
            end              
          end              
          self.new :headers => headers, :body => body
        end
                            
        # Implement enumerable                
        def each
          @body.each { |i| yield i }
        end                
                                
        private
        # Encodes row according to type
        def encode
          
        end
        
      end
    end
  end
end

lib/fusion_tables/data/table.rb
# Copyright (C) 2010 Simon Tokumine
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

module GData
  module Client
    class FusionTables < Base
      class Table
        attr_reader :headers, :id, :name                      
        
        # configures headers hash and sets up
        #
        # eg options: {:table_id => "x", :name => "y"}
        #
        def initialize client, options
          raise ArgumentError, "need ft client" if client.class != GData::Client::FusionTables 
          raise ArgumentError, "need table_id and name hash" if !options.has_key?(:name) || !options.has_key?(:table_id) 
          @client = client
          @id = options[:table_id]
          @name = options[:name]
        end

        # Sets up data types from google
        #
        def describe
          @client.execute "DESCRIBE #{@id}"
        end
                
        # Runs select and returns data obj
        #
        # Define columns and SQL conditions separatly
        # 
        # See http://code.google.com/apis/fusiontables/docs/developers_reference.html#Select 
        #
        # use columns=ROWID to select row ids
        #
        def select columns="*", conditions=nil
          @client.execute "SELECT #{columns} FROM #{@id} #{conditions}"
        end
        
        # Returns a count of rows. SQL conditions optional
        #
        # Note: handles odd FT response: when table has 0 rows, returns empty array.
        def count conditions=nil
          result = select("count()", conditions)
          result.empty? ? 0 : result.first[:"count()"].to_i          
        end
        
        
        # Outputs data to an array of concatenated INSERT SQL statements
        #
        # format should be:
        #
        # [{:col_1 => data, :col2 => data}, {:col_1 => data, :col2 => data}]
        #
        # Fields are escaped and formatted for FT based on type
        #
        def insert data
          data = [data] unless data.respond_to?(:to_ary)

          # encode values to insert
          data = encode data
          
          # Chunk up the data and send
          chunk = ""
          data.each_with_index do |d,i|            
            chunk << "INSERT INTO #{@id} (#{ d.keys.join(",") }) VALUES (#{ d.values.join(",") });"
            if (i+1) % 500 == 0 || (i+1) == data.size
              begin
                @client.sql_post(chunk)
                chunk = ""
              rescue => e
                raise "INSERT to table:#{@id} failed on row #{i} with #{e}"
              end  
            end  
          end
        end                

        # Runs update on row specified and return data obj
        def update row_id, data          
          data = encode([data]).first
          data = data.to_a.map{|x| x.join("=")}.join(", ")
          @client.execute "UPDATE #{@id} SET #{data} WHERE ROWID = '#{row_id}'"
        end
        
        # delete row
        def delete row_id
          @client.execute "DELETE FROM #{@id} WHERE rowid='#{row_id}'"
        end
        
        # delete all rows
        def truncate!
          @client.execute "DELETE FROM #{@id}"
        end
        
        def get_headers
          @headers ||= describe
        end                  
        
        def encode data
          data.inject([]) do |ar,h|
            ret = {}
            h.each do |key, value|              
              if value.nil?
                #empty string for nils
                ret["'#{key.to_s}'"] = "''"
              else
                ret["'#{key.to_s}'"] = case get_datatype(key)
                  when "number"   then  "#{value}"
                  when "datetime" then  "'#{value.strftime("%m-%d-%Y %H:%M:%S")}'"
                  else                  "'#{value.gsub(/\\/, '\&\&').gsub(/'/, "''")}'"                            
                end
              end
            end
            ar << ret
            ar      
          end
        end  
                        
        # 
        # Returns datatype of given column name
        #
        def get_datatype column_name
          get_headers
          
          @headers.each do |h|
            return h[:type] if h[:name] == column_name.to_s
          end            
          raise ArgumentError, "The column doesn't exist"
        end      
      end
    end
  end
end


lib/fusion_tables/client/fusion_tables.rb
# Copyright (C) 2010 Tom Verbeure, Simon Tokumine
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

module GData
  module Client
    class FusionTables < Base

      SERVICE_URL = "https://tables.googlelabs.com/api/query"
      DATATYPES  = %w(number string location datetime)
      
      def initialize(options = {})
        options[:clientlogin_service] ||= 'fusiontables'
        options[:headers] = { 'Content-Type' => 'application/x-www-form-urlencoded' }
        super(options)
      end

      def sql_encode(sql)
        "sql=" + CGI::escape(sql)
      end

      def sql_get(sql)
        resp = self.get(SERVICE_URL + "?" + sql_encode(sql))
      end

      def sql_post(sql)
        resp = self.post(SERVICE_URL, sql_encode(sql))
      end

      def sql_put(sql)
        resp = self.put(SERVICE_URL, sql_encode(sql))
      end
                  
      # Overrides auth_handler= so if the authentication changes,
      # the session cookie is cleared.
      def auth_handler=(handler)
        @session_cookie = nil
        return super(handler)
      end

      # Overrides make_request to handle 500 redirects with a session cookie.
      def make_request(method, url, body = '', retries = 10)
        begin
          response = super(method, url, body)
        rescue GData::Client::ServerError => e
          if e.response.status_code == 500 and retries > 0
            sleep_time = 11 - retries 
            sleep sleep_time # <= Fusion tables has rate of 5 calls per second. Be nice, get longer
            @session_cookie = e.response.headers['set-cookie']          
            return self.make_request(method, url, body, retries - 1)
          else
            return e.response
          end  
        end  
      end

      # Custom prepare_headers to include the session cookie if it exists
      def prepare_headers
        if @session_cookie
          @headers['cookie'] = @session_cookie
        end
        super
      end
    end
  end
end


examples/boris_bikes.rb
# This library creates a FT and posts data from the Boris Bikes API to it every minute
#
# Add photos to infowindow
# Add fusion table graphs to the inside of the infowindows too


require 'net/http'
require 'uri'
require 'rubygems'
require 'geo_ruby'
require 'fusion_tables'
require 'time'
require 'json'
require 'yaml'
require 'ap'
include GeoRuby
include SimpleFeatures

class Object
  def try(method, *args, &block)
    send(method, *args, &block)
  end
end

def color max, number_to_color, min=0, opacity=80
  color = ["FFFFB2", "FFFFB2", "FEB24C", "FD8D3C", "F03B20", "BD0026"]
  #color = %w(FEE0D2 FCBBA1 FC9272 FB6A4A EF3B2C CB181D A50F15 67000D)
  color.reverse!
  #color = ["FFFFCC", "D9F0A3", "ADDD8E", "78C679", "31A354", "31A354"] #<- greens
  chunk = (max-min)/color.size
  index = (number_to_color/chunk).floor
  "#{color[index]}#{opacity}"
end

def to_google(x,y)
  a = `echo "#{x} #{y}" | cs2cs + +init=epsg:4326 +to +init=epsg:3785 -f "%.12f"`
  a = a.split(" ")
  {:x => a[0], :y => a[1]}
end

def from_google(x,y)
  a = `echo "#{x} #{y}" | cs2cs + +init=epsg:3785 +to +init=epsg:4326 -f "%.12f"`
  a = a.split(" ")
  {:x => a[0], :y => a[1]}
end


def buffer(center_x, center_y, radius, quality = 4, precision = 12)
  points = []
  radians = Math::PI / 180
  
  coords = to_google(center_x, center_y)
  center_x = coords[:x].to_f
  center_y = coords[:y].to_f
  
  0.step(360, quality) do |i|
    x = center_x + (radius * Math.cos(i * radians))
    y = center_y + (radius * Math.sin(i * radians))
    coords = from_google(x,y)
    points << Point.from_x_y(round(coords[:x].to_f, precision), round(coords[:y].to_f, precision))
  end
  points
end


def round number, precision = 12
  (number * 10**precision).round.to_f / 10**precision
end  

# Configure settings
config = YAML::load_file(File.join(File.dirname(__FILE__), 'credentials.yml'))
DEFAULT_SRID = 4328


# Configure fusion tables
ft = GData::Client::FusionTables.new
ft.clientlogin(config["google_username"], config["google_password"])
table_name = "Boris Bikes"
cols = [
  {:name => 'name',         :type => 'string'},  
  {:name => 'created_at',   :type => 'datetime'},
  {:name => 'updated_at',   :type => 'datetime'},
  {:name => 'boris_id',     :type => 'number'},
  {:name => 'temporary',    :type => 'number'},
  {:name => 'installed',    :type => 'number'},
  {:name => 'locked',       :type => 'number'},
  {:name => 'nb_empty_docs',:type => 'number'},
  {:name => 'nb_bikes',     :type => 'number'},
  {:name => 'nb_docs',      :type => 'number'},
  {:name => 'image',        :type => 'string'},
  {:name => 'geom',         :type => 'location'},  
  {:name => 'geom_fill',    :type => 'string'},  
  {:name => 'geom_border',  :type => 'string'},  
]

# Create FT if it doesn't exist
tables = ft.show_tables
table  = tables.select{|t| t.name == table_name}.first
table  = ft.create_table(table_name, cols) if !table

while true do
  bikes = JSON.parse(Net::HTTP.get(URI.parse('http://borisapi.heroku.com/stations.json')))  
  
  # get largest bike rack to calibrate buffer
  max = 0
  bikes.each do |b|
    slots = b["nb_empty_docks"] + b["nb_bikes"]
    max = slots if slots > max
  end  
  
  # loop through data constructing fusion table date
  data = []
  max_radius = 150.0 #in meters
  buffer_chunk = max_radius / max
  
  bikes.each do |b|
    if b["lat"].to_f > 50 #ignore non geographic ones
      docs = (b["nb_bikes"] + b["nb_empty_docks"])    
      geom = Polygon.from_points [buffer(b["long"].to_f, b["lat"].to_f, docs*buffer_chunk)]    
      #geom = Point.from_x_y b["long"].to_f, b["lat"].to_f
      
      data << {
        "name"          => b["name"],
        "created_at"    => Time::parse(b["created_at"]),
        "updated_at"    => Time::parse(b["updated_at"]),
        "boris_id"      => b["id"],
        "temporary"     => (b["temporary"] ? 1 : 0), 
        "installed"     => (b["installed"] ? 1 : 0),
        "locked"        => (b["locked"]    ? 1 : 0),
        "nb_empty_docs" => b["nb_empty_docks"],
        "nb_bikes"      => b["nb_bikes"],
        "nb_docs"       => docs,
        "image"         => "",
        "geom"          => geom.as_kml,
        "geom_fill"     => color(max,b["nb_bikes"]),
        "geom_border"   => color(max,b["nb_bikes"],0,"FF"),
      }
      puts "packing data for #{b["name"]}"
    end  
  end       
  
  # get current number of rows ready to delete
  row_ids = table.select "ROWID"
  
  # put new data up
  puts "sending bikes to fusion tables..."
  table.insert data
  
  # remove old data
  puts "deleting old rows"
  row_ids.each do |id|
    table.delete id[:rowid]
  end
    
  # Be nice and wait
  puts "...done! sleeping..."
  sleep 500
end

